src/org/opendatakit/validate/FormValidator.java

```diff
@@ -324,28 +311,81 @@ boolean stepThroughEntireForm(FormEntryModel model) throws InvalidReferenceExcep
         return outcome;
     }
 
-    void validate(String path) {
+    public void validateAndExitWithErrorCode(String path) {
+        try {
+            validate(path);
+        } catch (Exception e) {
+            errors.error("\nException: ", e);
+            setError(true);
+        }
 
-    	File src = new File(path);
-    	if ( !src.exists() ) {
-    		setError(true);
-    		System.err.println("File: " + src.getAbsolutePath() + " does not exist.");
-    		return;
-    	}
+        if (inError) {
+            errors.error("\nResult: Invalid");
+            System.exit(1);
+        } else {
+            System.exit(0);
+        }
+    }
+
+  public void validate(String path) {
+        File src = new File(path);
+        if (!src.exists()) {
+            setError(true);
+            errors.error("File: " + src.getAbsolutePath() + " does not exist.");
+            return;
+        }
 
         FileInputStream fis = null;
         try {
             fis = new FileInputStream(src);
+            validate(fis);
+        } catch (FileNotFoundException e) {
+            setError(true);
+            errors.error("Please choose a file before attempting to validate.");
+            return;
+        } finally {
+            if (fis != null) {
+                try {
+                    fis.close();
+                } catch (IOException e) {
+                    // ignore
+                    e.printStackTrace();
+                }
+            }
+        }
 
-	        // validate well formed xml
-	        // System.out.println("Checking form...");
+    }
+
+
+
+    public void validateText(String xml) {
+        validate(new ByteArrayInputStream(xml.getBytes()));
+    }
+
+
+    public void validate(InputStream xmlSource) {
+
+            byte[] xformData;
+            try {
+                // first read the whole form into memory since this stream is going to be read twice
+                // first for xml validation
+                // second for xform validation
+                ByteArrayOutputStream baos = copyToByteArray(xmlSource);
```

Comment:
Hi thanks for the review,

This was not about performance improvement on my side.. I was simply trying to have one `validate()` method that takes in an `InputStream` and uses it to validate both xml Well-formedness and xform correctness with minimal code changes to the original file(or method) so as to minimize any bugs I could introduce. 

To use the data from the stream twice I had to copy it and create two more streams on demand. I thought the memory footprint was a little negligible since most forms are usually less than 1MB(I might be wrong?).

Since I was not sure whether the supplied input stream might support `mark()` and `reset()` I thought this is my best painless path. Am very willing to change the code if you any suggestions.